### PR TITLE
fix(ivy): align injection behavior between ivy and ViewEngine

### PR DESCRIPTION
**Pull request by mhevery**
this is a follow up from https://github.com/angular/angular/pull/27387#pullrequestreview-180717673 discussion.

@marclaval Please look into it.

/cc @alexzuza

### CODE REVIEW

**Comment by trotyl** [comment]
Should the type be `test(ivy)` as being a test-only change?

**Comment by JoostK** [comment]
@trotyl this is just for reference now to find any potential difference in behavior, nothing has been changed yet.

@marclaval before you investigate, I have an idea on what's going on, see https://github.com/angular/angular/pull/27387#discussion_r238411340 (copied below for reference):

> that looks like a bug in ViewEngine. ViewEngine only performs `@Host` checking during template compiles, but I'm not sure it takes directives into account. In Ivy, the `@Host` is checked at runtime in the node injector itself during resolving.
> 
> In your example, the `ControlNameDirective` is applied to the `<input controlName type="text">` element. That particular element is the directive's host, not the component itself. So, the `viewProvider` of the container is not actually in scope, hence Ivy throws.
> 
> I discovered one more scenario where ViewEngine is able to resolve an `@Host` decorated param where Ivy is not, namely when dynamically creating a component. My understanding has always been that the element from which the component is created dynamically acts as the host, but Ivy has proven that not to be true. And that makes sense, because even for dynamically created components you'll get an actual host element according to the created component's selector (or `ng-component` if it has no selector).
> 
> In ViewEngine, the `@Host` decorator is verified in the parent template where the host element originates from. For dynamically created components no such static host element exists so ViewEngine's template parser does not enforce in any way the `@Host` decorator. ViewEngine would happily traverse the full node injector tree to resolve an `@Host` token, which is not desirable.
> 
> So, in conclusion: yes, you may find that Ivy is unable to resolve `@Host` annotated providers where ViewEngine was able to, but that actually boils down to a limitation in ViewEngine's checks.

File: packages/core/test/render3/di_spec.ts [hunk 1]
```diff
@@ -2071,4 +2072,65 @@ describe('di', () => {
...
+      class GroupDirective {
+      }
+
+
+      @Directive({selector: '[controlName]'})
+      class ControlNameDirective {
+        // This does not seem to work. When code is generated instead of `ControlContainer` we are
+        // injecting `null`
+        // constructor(@Host() @SkipSelf() parent: ControlContainer) {}
+        constructor(@Host() @SkipSelf() @Inject(ControlContainer) parent: ControlContainer) {
```
Comment (mhevery) [inline packages/core/test/render3/di_spec.ts]: The `ControlNameDirective`'s  `@Host` is `<input controlName type="text">` hence it should not see `ControlContainer`. So I think this test is just wrong, and current Ivy behavior is correct.

Comment (alexzuza) [reply packages/core/test/render3/di_spec.ts]: I don't argue that `<input/>` element is a Host for that directive but:

- Open Angular docs example for `@Host` https://angular.io/api/core/Host#usage-notes
- Check it in ViewEngine https://stackblitz.com/edit/angular-n2894a?file=src%2Fapp%2Fapp.module.ts
- Check it in Ivy(_click Start button_) https://alexzuza.github.io/ivy-jit-preview//#hVRLbtswEN3rFAM1gORCtveOZTh1UqRAUwdod4YB0xItEaVIgaKUGIaO1UPkVh1Sn8qukaxEzrw38.ZDsSyXSsMJVhJPggodwD1TNNKsogF8E3mJlrc_GVG_0a6PQAroLwH8SJ5kXHKEPlRIfsiY1lQFsC61Za5zzaQgPIBHWWio4aBkBt6SiKTkRE0jqah367BORs6JPkiVfVHypaDq_ihIxqL_eR1uvG.A47hBDmO1MRqBH4dAqhNxUhSw1ilVP6mqWEQxTt2aTQVDq7PsO.WfoKAcz1LNwItSxuNx3Dm9AOir0XRXdE4kelCP2sCr1tTA4eQAcJkguNCKiWSzhRA2W5QHEEmBxtIk8pddc_0RLI04_EpkDeUHcA2VImpQzcimBJhOkQ.sAFFyPgPMQ61dp6yYGEWTvCxSfzcE3ZzwFoahvda70W0XKbUgIoChYiKwZ_IwTPpO_Pl.8TH75mQwrffcWc.n.0UjpXZwTs6y327fVDoYVU4UWsdRlnsBeipGX56VrFhMFfZoMwi6NX5NM7M2mH5nlc8vRg2frCF0uym7i_n0ArNomDGrcL5HTkNX01c9JpwlYsbpQbsNosV8FslXqUKXUw1ZkZg.2IC2YS5smBBUPf56.r4NXfSbhEhrk3THXeD02_ZsS15lOU69fq832XFM8ut9Ga7YRWO8.b.eopbBxRuIuMvb9M6y.4XY9M3jNSnOXq_NEVPkKmK22QBshRgnsIe.quZ6_qSMzYbYS6nx_ZAc.UjdGkHN24ReV_u_sC_8.u_IH036QK32nodb9xc

Comment (mhevery) [reply packages/core/test/render3/di_spec.ts]: I had further discussion with @kara and @pkozlowski-opensource  and we came to the conclusion that the current behavior is correct. The name is weird `@Host` which is where the confusion lies. A better name would be `@ParentView`. So this is a legitimate Regression which needs to be fixed.

**Comment by mhevery** [comment]
Resolved in https://github.com/angular/angular/pull/27646

**Pull request closed without merging**